android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/Utils.java

```diff
@@ -351,7 +329,59 @@ public static boolean checkIfUserHasOptedIn(Context context) {
    * @param context The context for the service.
    */
   public static void deleteCache(Context context) {
-    new ClearCacheDiscoveryServiceConnection().connect(context);
+    (new ServiceConnection(){
+      private Context mContext;
+
+      @Override
+      public void onServiceConnected(ComponentName className, IBinder service) {
+        // Get the service
+        UrlDeviceDiscoveryService.LocalBinder localBinder =
+            (UrlDeviceDiscoveryService.LocalBinder) service;
+        localBinder.getServiceInstance().clearCache();
+        mContext.unbindService(this);
+      }
+
+      @Override
+      public void onServiceDisconnected(ComponentName className) {
+      }
+
+      public void connect(Context context) {
+        mContext = context;
+        Intent intent = new Intent(mContext, UrlDeviceDiscoveryService.class);
+        mContext.startService(intent);
+        mContext.bindService(intent, this, Context.BIND_AUTO_CREATE);
+      }
+    }).connect(context);
+  }
+
+  /**
+   * Starts scanning with UrlDeviceDisoveryService.
+   * @param context The context for the service.
+   */
+  public static void startScan(Context context) {
+    (new ServiceConnection(){
```

Comment:
I'm fine with this as is, but I'm wondering if it would be worthwhile to subclass service connection so that any implementing subclass gets a localBinder and gets to do stuff with it and doesn't have to manage the connection.